### PR TITLE
Fixed token refresh callback

### PR DESCRIPTION
**Pull request by lucasm-iRonin**
Hi,
I've spotted 2 issues related to token refresh which I think I fixed in the following PR. I'm aware I can remove duplication but for now I just wanted to post it for initial review (to check if it's the right direction).

Basically I see 2 problems related to token refresh at the moment:

1. When token is refreshed and `fetchData` is disabled callback for `_routerBeforeEach` is never called - I believe it's the root of the problem in #87 issue.
2. When token needs to be refreshed, data fetching (if enabled) does not wait for the refresh request to complete. It results in sending request for data fetch with expired token.

I've fixed both problems here but I tested it only with my use case.
I've started using this library recently so please keep in mind I may not be familiar with all use cases that needs to be handled.

### CODE REVIEW

**Comment by websanova** [comment]
I'm not sure I understand your point in #1. If fetch data is disabled, nothing should be called. It would have to be called manually.

For 2 this is by design, is it related to this? https://github.com/websanova/vue-auth#token-refresh

**Comment by lucasm-iRonin** [comment]
@websanova as fair as I understand the code this is the before route callback. No matter whether there is fetch data enabled / disabled you have to call it to proceed the route transition. Otherwise you will see a blank page as it was reported in #87 (I expected this as well).

Why #2 issue is by design? I'm sorry but I don't get it. If you have token refreshes and data fetching built in then data should be fetched with token after it's refreshed not before (with the expired value). You can achieve it with callback as I show in my code.

**Comment by websanova** [comment]
For

1. Are you using auth.check or auth.ready anywhere?

2. Did you read the notes?

https://github.com/websanova/vue-auth/blob/master/README.md#token-refresh

**Comment by lucasm-iRonin** [comment]
@websanova thanks for quick reply.

Re 1.
Yes, I was using `auth.check` in the `before` hook in the default route (to check uf user is authenticated and I redirect him to default authenticated route) when I encountered this problem.

Re 2.
I've read them but it's not clear to me how should I solve this problem with current infrastructure. Should I disabled refresh token and add custom global `before` hook which should refresh token manually if it's expired? 
Also there is no easy access to $auth service in before hooks. Currently I'm using this way:

```javascript
let auth = Vue.prototype.$auth;
```

Could you explain why previous strategy was removed in current version? "It was removed because of the overhead since it requires a base64 encode/decode". If you keep `expired_at` as part of the token value then you compare can compare if before each request without doing any base64 decoding / encoding. Do I miss anything?

I think this strategy has less overhead than refreshing the token on every page refresh.

I'm coming from ember land where I used really great [authentication addon](https://github.com/simplabs/ember-simple-auth) for a long time. I recommend you to check it to see how they deal with all those issues.

**Comment by websanova** [comment]
Do you have some code? Y
The cue auth install and tour intercept. Also versions you are using?

**Comment by websanova** [comment]
Just an update on this.

1. Probably would need a simple example here of your install and usage.

2. Maybe we need to define the issue more clearly to make sure we are on the same page. Maybe we cna define the issue outside of the plugin. Do you mean that; two requests are made. The first request issues a new token before the second request hits the server. The second request however has the old token so it get's locked out with bad token. In this case it's a refresh/user call?

**Comment by lucasm-iRonin** [comment]
@websanova sorry for reversed order.

Re 2
Yes this is exactly what I hope to fix in this PR: Two requests are made - one for token refresh, one for fetching user at the same time.
Because fetching user request is fired at the same time as token refresh, the user fetch is using the expired token.

Re 1
I will try to give you more detailed description of how we use this addon so you can have better understanding of the use case.
Basically are building Chrome Extension and when user opens the extension we load the current user using $auth service and if user is logged in we redirect him to one of the authenticated pages.
If token is expired, we need to refresh it, then fetch current user data and redirect him to one of the authenticated pages.
We need to load the current user because there is info about his latest activity which is used in the redirect.

On the backend we are using Rails with [Doorkeeper gem](https://github.com/doorkeeper-gem/doorkeeper).
Response for the token (login) and refresh token looks like this:

```json
"access_token": "xxx"
"expires_in": n
"refresh_token": "yyy"
"token_type": "bearer"
"user_id": 123456789
```

Now in order to adapt `vue-auth` for our use case we had to configure with the options (and use version from the PR) - 
[Gist](https://gist.github.com/lucasm-iRonin/8ef4b20a27d68d6fc6ed8e13cb1e96a7#file-configuration-js).

We also use custom `bearer` driver because we store token as serialized `JSON` object - [Gist](https://gist.github.com/lucasm-iRonin/8ef4b20a27d68d6fc6ed8e13cb1e96a7#file-bearer-js).

We need all this to be able to persist more info about the token which is required during user fetch and checking token expiry.
As you see we had to override some code which is not easy because some methods are private or does not allow to call `super`.

In the default route we check whether the user is authenticated and redirect him to one of the authenticated routes:

```javascript
  return [
    {
      path: '/',
      beforeEnter: (to, from, next) => {
        // NOTE:
        // Dirty workaround to get $auth service
        // when we don't have access to Vue instance.
        const auth = Vue.prototype.$auth;

        // User not logged in
        if(!auth.check()) {
          return next('/login');
        }

        const currentUser = auth.user();
        const activityId = currentUser['get-some-latest-activity'];
        if(activityId) {
          return next(`/${activityId}/all`);
        } else {
		  // Do something else
        }
      }
    }
	...
	other routes
  ];
```

I hope this helps.

**Comment by websanova** [comment]
OK, so for Re2. As far as I understand. A refresh (meaning new token) send
during some request can happen at any request. We are looking at specific
scenario here refresh/user. But actually any two requests can have this
issue.

So wouldn't present a larger architectural issue, not just this one case?

**Comment by lucasm-iRonin** [comment]
> A refresh (meaning new token) send during some request can happen at any request. We are looking at specific scenario here refresh/user. But actually any two requests can have this issue.

@websanova you are right, token can expire at any moment and it can be invalid before any request other request to the API. In our case (because it is chrome extension) the refresh request will be the first request since we fetch current user before entering any route.

What's your state with this PR? If you don't want to merge it just let me know (I will keep my fork for my app purpose - I just need to know where we state with this). There was another problem I mentioned here - the one from #87 (when token is expired, callback on the route is not called and redirect seems to be not working - after using the code from this PR it was working for me).

Basically in my opinion if this extension provides option to fetch data before entering the route, it should do it after refresh request is performed. I know it's broader problem (like you say any request in the app can have this issue) but we can try to solve it by using [vue-resource#interceptors](https://github.com/pagekit/vue-resource/blob/develop/docs/http.md#interceptors) in the future.

**Comment by websanova** [comment]
No, I can't merge this.

It doesn't solve the token problem at all since a new token can be
delivered in any request.

I guess you are strictly sending a token during a refresh call once at the
beginning. The scenario is already addressed by disabling the user data
fetch call. So you can call it manually in the refresh success callback.



On Feb 17, 2017 3:00 PM, "Lucas Mendelowski" <notifications@github.com>
wrote:

A refresh (meaning new token) send during some request can happen at any
request. We are looking at specific scenario here refresh/user. But
actually any two requests can have this issue.

@websanova <https://github.com/websanova> you are right, token can expire
at any moment and it can be invalid before any request other request to the
API. In our case (because it is chrome extension) the refresh request will
be the first request since we fetch current user before entering any route.

What's your state with this PR? If you don't want to merge it just let me
know (I will keep my fork for my app purpose - I just need to know where we
state with this). There was another problem I mentioned here - the one from
#87 <https://github.com/websanova/vue-auth/issues/87> (when token is
expired, callback on the route is not called and redirect seems to be not
working - after using the code from this PR it was working for me).

Basically in my opinion if this extension provides option to fetch data
before entering the route, it should do it after refresh request is
performed. I know it's broader problem (like you say any request in the app
can have this issue) but we can try to solve it by using
vue-resource#interceptors
<https://github.com/pagekit/vue-resource/blob/develop/docs/http.md#interceptors>
in the future.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/websanova/vue-auth/pull/97#issuecomment-280581228>, or mute
the thread
